### PR TITLE
CSharp: SWIG 2.x and 3.x compatibility

### PR DESCRIPTION
**Pull request by miurahr**
## What does this PR do?

- Switch code chunks using SWIG pre-processor and variable SWIG_VERSION
  in interface file to support backward compatibility for 1.3.40.

- Add -DSWIG2_CSHARP to swig command line for backward compatibility for
  2.0.x, that 3.0.x don't define using System directive in intermediate
  csharp codes.

- Tests for CSharp binding  on Travis(Trusty) and AppVeyor(Win32/64) CI.

- Drop installation of SWIG-1.3.40 and enable make test for csharp on Vagrant environment.

## What are related issues/pull requests?

Core modifications are discussed in
  http://osgeo-org.1560.x6.nabble.com/gdal-dev-CSharp-bindings-question-tt5207528.html#a5210247
  in 3 years ago

## Tasklist

 - [x] Add test
 - [x] Review
 - [x] Adjust for comments
 - [ ] All CI builds and checks have passed

### CODE REVIEW

**Comment by rouault** [comment]
@szekerest Could you review this ? Note that this will probably mean that you need to upgrade your gisinternals builds to swig >= 2

**Comment by rouault** [comment]
There's an issue with the MSVC 2015 x64 config: https://ci.appveyor.com/project/OSGeo/gdal/build/1.0.6638/job/dyqxj75q5yb7owkr

**Comment by szekerest** [comment]
@rouault That doesn't seem to be related to the SWIG version. I've re-enabled csharp tests at gisinternals and it works correctly.
The error at appveyor is looking like a BadImageFormatException which happens most likely when not all of the dll-s are compiled for x64 that are being loaded at run time.

**Comment by miurahr** [comment]
@rouault @szekerest I now understand x64 issue in appveyor script which is missing WIN64 argument.
Also I fixed Mono/Linux compilation problem in GNUmakefile which is missing libgdal linking, (but   there is a space to improve).

Then now there are csharp binding tests both mono/trusty and .Net/Windows that will pass.

File: appveyor.yml [hunk 1]
```diff
@@ -135,6 +135,9 @@ install:
       cd $env:APPVEYOR_BUILD_FOLDER\gdal\swig\python
       exec { python setup.py build }
       exec { python setup.py install }
+      cd $env:APPVEYOR_BUILD_FOLDER\gdal\swig\csharp
+      exec { nmake /f makefile.vc interface}
+      exec { nmake /f makefile.vc all $env:WIN64_ARG}
```
Comment (miurahr) [inline appveyor.yml]: environment variable WIN64_ARG defines WIN64 which affect csc option for architecture.

File: gdal/ci/travis/trusty_clang/install.sh [hunk 2]
```diff
@@ -49,6 +49,10 @@ cd swig/perl
 make generate
 make
 cd ../..
+cd swig/csharp
+make generate
+make
+cd ../..
```
Comment (miurahr) [inline gdal/ci/travis/trusty_clang/install.sh]: we can now add `make test`.

File: gdal/swig/include/csharp/swig_csharp_extensions.i [hunk 3]
```diff
@@ -9,11 +9,14 @@
  *
 */
 
+#if (SWIG_VERSION < 0x020000)
+// On SWIG 1.3.x
```
Comment (miurahr) [inline gdal/swig/include/csharp/swig_csharp_extensions.i]: redundant comment

File: gdal/swig/csharp/GNUmakefile [hunk 4]
```diff
@@ -7,10 +7,10 @@ all: build
 BINDING = csharp
 include ../SWIGmake.base
 
-SWIGARGS = -Wall -I../include -I../include/$(BINDING)
+SWIGARGS = -Wall -I../include -I../include/$(BINDING) -DSWIG2_CSHARP
 
 LINK = $(LD_SHARED)
-LINK_EXTRAFLAGS = 
+LINK_EXTRAFLAGS =  $(GDAL_SLIB_LINK)
 OBJ_EXT = o
```
Comment (miurahr) [inline gdal/swig/csharp/GNUmakefile]: Are there better place to add $(GDAL_SLIB_LINK) ?

**Comment by miurahr** [comment]
Here is missing place to add `-DSWIG2_CSHARP` in  `gdal/swig/csharp/mkinterface.sh`

File: gdal/swig/csharp/GNUmakefile [hunk 5]
```diff
@@ -7,10 +7,11 @@ all: build
 BINDING = csharp
 include ../SWIGmake.base
 
+SWIG = $(SWIG) -DSWIG2_CSHARP
```
Comment (rouault) [inline gdal/swig/csharp/GNUmakefile]: Probably that this should be the :=  operator

**Comment by rouault** [comment]
It currently breaks the trusty_clang config. Otherwise @szekerest if you're happy, I'm too

**Comment by szekerest** [comment]
At the moment I don't have such environment to test linux build issues. It would be more compelling to fix this item before merging if possible. Otherwise the changes look fine for me.

**Pull request merged**